feat: Adding support for Cassandra and Scylla

fruch opened the pull request:
This add the support for those cassandra based dbs
and their drivers, cassandra-driver, scylla-driver

Ref: https://cassandra.apache.org/
Ref: https://www.scylladb.com/
Ref: https://pypi.org/project/cassandra-driver/
Ref: https://pypi.org/project/scylla-driver/

fruch commented:
Hi @tillahoffmann and @SergeyPirogov, pretty please for approving the workflow to start ?

FYI, it can be eased a bit, the approval part:
https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#configuring-required-approval-for-workflows-from-public-forks

yakimka commented:
@fruch hello. Can you update this PR?

tillahoffmann commented:
Thanks for the additions and apologies for the slow reply. A few things to get this integrated:

- Rebase on master.
- Regenerate requirement files.
- Add `cassandra` and `scylla` to the `test-components` in `.github/workflows/main.yml`.

fruch commented:
> Thanks for the additions and apologies for the slow reply. A few things to get this integrated:
> 
> * Rebase on master.
> * Regenerate requirement files.
> * Add `cassandra` and `scylla` to the `test-components` in `.github/workflows/main.yml`.

I've rebased and added the test to CI

but the building of requirements is failing:
```
There are incompatible versions in the resolved dependencies:
  protobuf<4.0.0dev (from google-cloud-pubsub==1.7.2->testcontainers===dev->-r requirements.in (line 1))
  protobuf<5.0.0dev,>=3.20.1 (from google-api-core[grpc]==2.10.1->google-cloud-pubsub==1.7.2->testcontainers===dev->-r requirements.in (line 1))
  protobuf<5.0.0dev,>=3.15.0 (from googleapis-common-protos[grpc]==1.56.4->google-api-core[grpc]==2.10.1->google-cloud-pubsub==1.7.2->testcontainers===dev->-r requirements.in (line 1))
  protobuf>=4.21.3 (from grpcio-status==1.49.0->google-api-core[grpc]==2.10.1->google-cloud-pubsub==1.7.2->testcontainers===dev->-r requirements.in (line 1))
```

I think pinning of google-cloud-pubsub is problematic, and google packages are conflicting each other.
I'm trying to unpin it, and regenerate (but it quite a long process, meanwhile it worked for python3.7)

fruch commented:
@yakimka @tillahoffmann waiting for approval for running the CI

tillahoffmann commented:
We may have to get #209 merged first to get around the google dependency issue.

tillahoffmann commented:
I have a relatively strong preference to remove the cache for two reasons. First, it's more code to maintain. Second, I don't think the caching will help in most real-world scenarios. 

I appreciate that runtime is different across different machines, but getting a single image takes about 4ms on my machine. Using the cache, getting a single image takes about 300ms. It'll of course be quicker the next time images are fetched, but we'd have to use at least 75 containers per test suite to make the cache worthwhile. To have a delay of one second over the cached version, we'd have to run about 325 containers. I'd argue that for a test suite that runs more than 300 docker containers a delay of one second is negligible. But let me know if I misunderstood/miscalculated.

fruch commented:
> I have a relatively strong preference to remove the cache for two reasons. First, it's more code to maintain. Second, I don't think the caching will help in most real-world scenarios. 
> 
> I appreciate that runtime is different across different machines, but getting a single image takes about 4ms on my machine. Using the cache, getting a single image takes about 300ms. It'll of course be quicker the next time images are fetched, but we'd have to use at least 75 containers per test suite to make the cache worthwhile. To have a delay of one second over the cached version, we'd have to run about 325 containers. I'd argue that for a test suite that runs more than 300 docker containers a delay of one second is negligible. But let me know if I misunderstood/miscalculated.

Didn't understand where you are going with the calculation here, for me getting a docker image out of dockerhub can take a minute or two. I was more talking about pip-compile that in case of conflicts can download all the versions of a package to try solving it, (I know it, cause I'm using it in in one of my projects in my day job),

And as you said I don't think it's worth while for a library to maintain such a lock file. even that tool like poetry for example are doing just that. (also can be slow and annoying)

tillahoffmann commented:
Sorry, I posted on the wrong thread. 🤦‍♂️

fruch commented:
> We may have to get #209 merged first to get around the google dependency issue.

@tillahoffmann would it be o.k. to take the code changes from #209, and test it in tandam ?

spicy-sauce commented:
Hi,
What's the status of this PR? Do you plan to merge it to the master anytime soon?

@fruch I see there are conflicts that should be resolved, is it something that you should handle?

Thanks,
Yossi

fruch commented:
> Hi,
> What's the status of this PR? Do you plan to merge it to the master anytime soon?
> 
> @fruch I see there are conflicts that should be resolved, is it something that you should handle?

I guess I could, I've almost forgotten about this one, it's almost a whole year since I've opened this one.

> 
> Thanks,
> Yossi

spicy-sauce commented:
> I guess I could, I've almost forgotten about this one, it's almost a whole year since I've opened this one.
Thanks!

@tillahoffmann can we please push it once @fruch resolves the conflicts? This seems like a great addition to testcontainers!

fruch commented:
> > I guess I could, I've almost forgotten about this one, it's almost a whole year since I've opened this one.
> 
> Thanks!
> 
> @tillahoffmann can we please push it once @fruch resolves the conflicts? This seems like a great addition to testcontainers! 
> 
> 

BTW @spicy-sauce can you share a bit what do you need it for ? Cassandra or Scylla ? And which type of testing it's gonna be part of ?

spicy-sauce commented:
> BTW @spicy-sauce can you share a bit what do you need it for ? Cassandra or Scylla ? And which type of testing it's gonna be part of ?

At this point I need it for Cassandra as I'm going to implement such a [block ](https://github.com/datayoga-io/datayoga/issues/124 )in [DataYoga](https://datayoga-io.github.io/datayoga) and would love to use it for the testing part.

Read more here:
https://datayoga-io.github.io/datayoga/

fruch commented:
> > BTW @spicy-sauce can you share a bit what do you need it for ? Cassandra or Scylla ? And which type of testing it's gonna be part of ?
> 
> At this point I need it for Cassandra as I'm going to implement such a [block ](https://github.com/datayoga-io/datayoga/issues/124)in [DataYoga](https://github.com/datayoga-io/datayoga) and would love to use it for the testing part.
> 
> Read more here: https://datayoga-io.github.io/datayoga/

looks interesting, ping me when you have it running with Cassandra or Scylla,  maybe adding it in our https://university.scylladb.com/

anyhow I'm the maintainer of python driver fork
https://github.com/scylladb/python-driver

if you have question you can reach me at fruch@scylladb.com

fruch commented:
@tillahoffmann #209 didn't seemed to be merged

Anyhow I've rebased again, and rebuild the requirements files again.

Can now it be run in CI and merged ?

fruch commented:
@tillahoffmann looks like I've missed updating the requriements files for 3.7, now it should run o.k.

spicy-sauce commented:
Guys, can we please push this one? It's really great addition and we would like to avoid another conflicts.
Thanks a lot!!!

fruch commented:
@tillahoffmann @yakimka it's now passes all the tests...

tillahoffmann commented on testcontainers/cassandra.py:
```diff
@@ -0,0 +1,46 @@
+
+    Example
+    -------
+    ::
```
Sorry, one more. Can we add `.. doctest::` here (and the other container) to make things consistent with #268.

fruch replied:
done.

we are 10 days from this PR birthday, can we have a celebration ?

spicy-sauce replied:
@tillahoffmann Can we please finalize this?

spicy-sauce replied:
@fruch While we are waiting for it, seems that both new databases should be added to `README.rst` as well.

spicy-sauce commented:
@fruch While we are waiting for it, can you please add the new containers to `README.rst`?
@tillahoffmann I see RedisContainer is also missing.

fruch commented:
> @fruch While we are waiting for it, can you please add the new containers to `README.rst`?
> @tillahoffmann I see RedisContainer is also missing.

I first need to fix the doctests, f**king white spaces seems to be breaking it.

spicy-sauce commented:
@fruch this one failed:
testcontainers-python / build (3.9, cassandra.py)

https://github.com/testcontainers/testcontainers-python/actions/runs/3788545927/jobs/6442220582

Thanks!

fruch commented:
> @fruch this one failed:
> testcontainers-python / build (3.9, cassandra.py)
> 
> https://github.com/testcontainers/testcontainers-python/actions/runs/3788545927/jobs/6442220582
> 
> Thanks!

Well it's a bit hard to debug the issue, since those doctests are not collecting the docker images logs...

Kind of pointless to have the same tests written, and in the doctests, it's just repeating the same tests with any pytest facilities you can write to collect information.

Also I can rerun the action on my own, so can even gather statistics on the issue.

tillahoffmann commented:
> Kind of pointless to have the same tests written, and in the doctests, it's just repeating the same tests with any pytest facilities you can write to collect information.

The main point of the doctests is to ensure the documentation remains consistent with any changes that might be made to the interface rather than testing the code for correctness.

I've rerun the tests and everything is passing. One more rebase, and we should be able to merge this one. Thank you for the patience.

fruch commented:
> > Kind of pointless to have the same tests written, and in the doctests, it's just repeating the same tests with any pytest facilities you can write to collect information.
> 
> The main point of the doctests is to ensure the documentation remains consistent with any changes that might be made to the interface rather than testing the code for correctness.
> 
> I've rerun the tests and everything is passing. One more rebase, and we should be able to merge this one. Thank you for the patience.

fixed merge conflicts

tillahoffmann commented on setup.py:
```diff
@@ -73,6 +73,8 @@
         'azurite': ['azure-storage-blob'],
         'opensearch': ['opensearch-py'],
+        'cassandra': ['cassandra-driver'],
+        'scylla': ['scylla-driver'],
```
It looks like this dependency is not used?

fruch replied:
They differ in implementation, logging of both products isn't identical, also configuration isn't identical.

As for the drivers we need them both, depends on what one wants to test with

tillahoffmann replied:
It looks like the `ScyllaContainer` directly uses the Cassandra client here.

https://github.com/testcontainers/testcontainers-python/pull/167/files#diff-01edf1d9738ed56417c0881509e1db05ac47b51df3015ed2cee7bc5ed54a2c8dR44

fruch replied:
Both drivers use the same namespace, since scylla one is a drop-in replacement for the cassandra one.

I still think users should be able to choose between them in a documented way.

tillahoffmann replied:
Unfortunately, while `cassandra-driver` and `scylla-driver` have different pypi names, they have the same package name. They should thus not be installed together because the behavior is non-deterministic: Which package is used depends on the installation order. Consider the following scripts.

```python
# show.py: Shows the __init__.py location and md5 digest of the file.
import cassandra
import hashlib

with open(cassandra.__file__, "rb") as fp:
    digest = hashlib.md5(fp.read()).hexdigest()

print(cassandra.__file__, digest)
```

```bash
# Install the packages in different orders and compare the hashes.
echo Cassandra only.
pip uninstall -y scylla-driver cassandra-driver > /dev/null 2> /dev/null
pip install cassandra-driver > /dev/null
python playground/show.py

echo Scylla only.
pip uninstall -y scylla-driver cassandra-driver > /dev/null 2> /dev/null
pip install scylla-driver > /dev/null
python playground/show.py

echo Cassandra and Scylla.
pip uninstall -y scylla-driver cassandra-driver > /dev/null 2> /dev/null
pip install cassandra-driver > /dev/null
pip install scylla-driver > /dev/null
python playground/show.py

echo Scylla and Cassandra.
pip uninstall -y scylla-driver cassandra-driver > /dev/null 2> /dev/null
pip install scylla-driver > /dev/null
pip install cassandra-driver > /dev/null
python playground/show.py
```

```
# output
Cassandra only.
/usr/local/python/3.10.4/lib/python3.10/site-packages/cassandra/__init__.py ff2ebcf0b1df99a66633b96df60eccd8
Scylla only.
/usr/local/python/3.10.4/lib/python3.10/site-packages/cassandra/__init__.py cf43e7f36e8479d6403aec8d128e0990
Cassandra and Scylla.
/usr/local/python/3.10.4/lib/python3.10/site-packages/cassandra/__init__.py cf43e7f36e8479d6403aec8d128e0990
Scylla and Cassandra.
/usr/local/python/3.10.4/lib/python3.10/site-packages/cassandra/__init__.py ff2ebcf0b1df99a66633b96df60eccd8
```

I appreciate end users may want to install different packages depending on their use case, but we can unfortunately not install both in our CI because only one of them will be used. It looks like Scylla is a derivative of Cassandra. I'd suggest dropping the `scylla-driver` and `ScyllaContainer` and adding a note to the docstring that scylla can be used by modifying the image name (and optionally installing a different driver).

fruch replied:
Their docker image still behaves different, config in different ways, have different command line arguments

And different logs, current implementation uses logs to figure it's ready.

For CI we can install only one of the drivers.

If I need to implement my own code to use Scylla, then this PR doesn't have value for me, the whole purpose was to ease the testing story of both cassandra and scylla users.

tillahoffmann replied:
I appreciate that the two containers are different, but we don't need two separate classes because the functionality is the same in this library. E.g., the same log message is being checked [here](https://github.com/testcontainers/testcontainers-python/pull/167/files#diff-1f57e9d177cbc6ae423f97462c5869afb98e6343e2e21aea3f9674a6ba09dd4fR32) and [here](https://github.com/testcontainers/testcontainers-python/pull/167/files#diff-01edf1d9738ed56417c0881509e1db05ac47b51df3015ed2cee7bc5ed54a2c8dR33).

fruch replied:
If one class would inherit for the other, it would o.k. ?

It does differ in command line argument

tillahoffmann replied:
Yes, the command line arguments may differ, but the implementation does not make use of different command line arguments. There is no functional difference between the two implementations that I can see other than the name of the image (see [here](https://www.diffchecker.com/vrdoCzA5/)).

spicy-sauce replied:
seems like we are almost there, can we give a last push and finish with it, @fruch? seems like very useful to me.

spicy-sauce replied:
are we going to do something with it or abandoning this?

tillahoffmann reviewed (commented):
Do I understand correctly that the implementations of `ScyllaContainer` and `CassandraContainer` only differ in terms  of the image and class names? If so, let's drop one of them as we've done for other containers, e.g., #265.

guy9 commented:
Is there anything preventing this from being merged?

alexanderankin commented:
Currently it doesn't make too much of a difference what is merged or not, because there has been no activity in the repository from either individuals with Pypi publish credentials. E.g. see #405. That said next week I think we are going to start the process according to pep 541 to reclaim the Pypi name and start releasing and merging changes again

spicy-sauce commented:
> Is there anything preventing this from being merged?

In the meanwhile, I'd solve the current conflicts.

SergeyPirogov commented:
There is a issue with relases because I live in Kyiv and we are fighting with Russsia, so there is no time for the project

alexanderankin commented:
@SergeyPirogov that very important and I wish for victory for Ukraine of course.

can we help with releases in the meantime? My username is daveankin on Pypi and if you add me as owner I will work with currently available maintainers to update this library and Pypi, without removing you from the project at all. I hope soon in the future you will have some time for the project and return but until then, please consider adding myself or someone from testcontainers/atomicjar for the meantime.

Thank you!

guy9 commented:
Also interested in this and happy to help

alexanderankin commented:
Maybe I can try to merge or rebase this branch locally or test just these
changes locally


since technically this is just maintenance we have to do to release after
merging

On Tue, Jan 23, 2024, 8:17 AM David Ankin ***@***.***> wrote:

> Latest is on this branch:
>
> * Merged testcontainers-core back with testcontainers like the other
> modules
> * Fixed tests on Ubuntu
> * Fixed docs/doc tests after moving modules/on Ubuntu
> * Fixed npipe bug on windows (still wondering about the right answer there
> in #415)
> * Wasn't able to check the whole windows test suite yet, where I left off
>
>
> https://github.com/testcontainers/testcontainers-python/compare/main...alexanderankin:testcontainers-python:poetry-refactor-changes-to-publish
>
> On Tue, Jan 23, 2024, 6:53 AM Guy Shtub ***@***.***> wrote:
>
>> Also interested in this and happy to help
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/testcontainers/testcontainers-python/pull/167#issuecomment-1905879700>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACECGJBGE7SUCW5L5D67OKLYP6QDNAVCNFSM5LAIYIL2U5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TCOJQGU4DOOJXGAYA>
>> .
>> You are receiving this because you commented.Message ID:
>> ***@***.***>
>>
>

fruch commented:
> hi @fruch,
> 
> i have backed up your commit as it exists today in here: https://github.com/testcontainers/testcontainers-python/commit/dd6520d95dc432e25ee2ef56a332f9d4b3d12a90 - please confirm its okay to rebase and push some commits to your branch to adjust to the new structure (#408)?
> 
> i see that you have enabled `Maintainers are allowed to edit this pull request.`, so thank you for that, but i am in a habit of asking before i force push, so I appreciate your timely response (or we will figure out something to do for a cassandra implementation based off of this one, i suppose)
> 
> thanks,
> David

Go right ahead, ping me you are done, I would give it a look (can do that also after it merged)

alexanderankin commented:
i dont think we actually want to merge this when it references an abandoned docker image for cassandra, so ive made some adjustments here - https://github.com/fruch/testcontainers-python/compare/cassandra_scylla_support...testcontainers:testcontainers-python:cassandra_scylla_support_image_improvement

alexanderankin commented:
rebased so this is ready to be merged

linking related - https://github.com/testcontainers/testcontainers-java/pull/8002/files
